master/internal/api_user.go

```diff
@@ -427,19 +431,23 @@ func (a *apiServer) PatchUser(
 
 	var ug *model.AgentUserGroup
 	if aug := req.User.AgentUserGroup; aug != nil {
-		if err = validateProtoAgentUserGroup(aug); err != nil {
-			return nil, err
-		}
-		ug = &model.AgentUserGroup{
-			UID:   int(*req.User.AgentUserGroup.AgentUid),
-			GID:   int(*req.User.AgentUserGroup.AgentGid),
-			User:  *req.User.AgentUserGroup.AgentUser,
-			Group: *req.User.AgentUserGroup.AgentGroup,
-		}
+		if checkEmptyProtoAgentUserGroup(req.User.AgentUserGroup) {
+			ug = &model.AgentUserGroup{}
+		} else {
+			if err = validateProtoAgentUserGroup(aug); err != nil {
+				return nil, err
+			}
+			ug = &model.AgentUserGroup{
+				UID:   int(*req.User.AgentUserGroup.AgentUid),
+				GID:   int(*req.User.AgentUserGroup.AgentGid),
+				User:  *req.User.AgentUserGroup.AgentUser,
+				Group: *req.User.AgentUserGroup.AgentGroup,
+			}
 
-		if err = user.AuthZProvider.Get().
-			CanSetUsersAgentUserGroup(ctx, *curUser, targetUser, *ug); err != nil {
-			return nil, status.Error(codes.PermissionDenied, err.Error())
+			if err = user.AuthZProvider.Get().
+				CanSetUsersAgentUserGroup(ctx, *curUser, targetUser, *ug); err != nil {
+				return nil, status.Error(codes.PermissionDenied, err.Error())
+			}
```

Comment:
I think the first "remove agent user group" branch should also check for `CanSetAgentUserGroup` permission.
But if you take the AgentUserGroupFromProto this entire code should be easier:

```
ug, err := model.AgentUserGroupFromProto(req.User.AgentUserGroup)
if err != nil {
  return nil, err
}
if err = user.AuthZProvider.Get().
			CanSetUsersAgentUserGroup(ctx, *curUser, targetUser, *ug); err != nil {
				CanSetUsersAgentUserGroup(ctx, *curUser, targetUser, *ug); err != nil {
			return nil, status.Error(codes.PermissionDenied, err.Error())
				return nil, status.Error(codes.PermissionDenied, err.Error())
			}
...